File: api_gateway/src/ui/controller/purchase/shipment_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/purchase"
	service "github.com/serp/api-gateway/src/core/service/purchase"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type ShipmentController struct {
	shipmentService service.IShipmentService
}

func (s *ShipmentController) CreateShipment(ctx *gin.Context) {
	var req request.CreateShipmentRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.CreateShipment(ctx.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) UpdateShipment(ctx *gin.Context) {
	shipmentId := ctx.Param("shipmentId")
	if shipmentId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	var req request.UpdateShipmentRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.UpdateShipment(ctx.Request.Context(), shipmentId, &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) DeleteShipment(ctx *gin.Context) {
	shipmentId := ctx.Param("shipmentId")
	if shipmentId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.DeleteShipment(ctx.Request.Context(), shipmentId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) GetShipment(ctx *gin.Context) {
	shipmentId := ctx.Param("shipmentId")
	if shipmentId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.GetShipment(ctx.Request.Context(), shipmentId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) GetShipmentsByOrderId(ctx *gin.Context) {
	orderId := ctx.Param("orderId")
	if orderId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.GetShipmentsByOrderId(ctx.Request.Context(), orderId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) ImportShipment(ctx *gin.Context) {
	shipmentId := ctx.Param("shipmentId")
	if shipmentId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.ImportShipment(ctx.Request.Context(), shipmentId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) AddItemToShipment(ctx *gin.Context) {
	shipmentId := ctx.Param("shipmentId")
	if shipmentId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	var req request.AddItemToShipmentRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.AddItemToShipment(ctx.Request.Context(), shipmentId, &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) UpdateItemInShipment(ctx *gin.Context) {
	shipmentId := ctx.Param("shipmentId")
	itemId := ctx.Param("itemId")
	if shipmentId == "" || itemId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	var req request.UpdateItemInShipmentRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.UpdateItemInShipment(ctx.Request.Context(), shipmentId, itemId, &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) DeleteItemFromShipment(ctx *gin.Context) {
	shipmentId := ctx.Param("shipmentId")
	itemId := ctx.Param("itemId")
	if shipmentId == "" || itemId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.DeleteItemFromShipment(ctx.Request.Context(), shipmentId, itemId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (s *ShipmentController) UpdateShipmentFacility(ctx *gin.Context) {
	shipmentId := ctx.Param("shipmentId")
	if shipmentId == "" {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	var req request.UpdateShipmentFacilityRequest
	if err := ctx.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	res, err := s.shipmentService.UpdateShipmentFacility(ctx.Request.Context(), shipmentId, &req)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func NewShipmentController(shipmentService service.IShipmentService) *ShipmentController {
	return &ShipmentController{
		shipmentService: shipmentService,
	}
}

File: purchase_service/src/main/java/serp/project/purchase_service/controller/ShipmentController.java
package serp.project.purchase_service.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import serp.project.purchase_service.dto.request.*;
import serp.project.purchase_service.dto.response.GeneralResponse;
import serp.project.purchase_service.dto.response.ShipmentDetailResponse;
import serp.project.purchase_service.entity.InventoryItemDetailEntity;
import serp.project.purchase_service.entity.ShipmentEntity;
import serp.project.purchase_service.exception.AppErrorCode;
import serp.project.purchase_service.exception.AppException;
import serp.project.purchase_service.service.InventoryItemDetailService;
import serp.project.purchase_service.service.ShipmentService;
import serp.project.purchase_service.util.AuthUtils;

import java.util.List;

@RestController
@RequestMapping("/purchase-service/api/v1/shipment")
@RequiredArgsConstructor
@Validated
@Slf4j
public class ShipmentController {

        private final ShipmentService shipmentService;
        private final InventoryItemDetailService inventoryItemDetailService;
        private final AuthUtils authUtils;

        @PostMapping("/create")
        public ResponseEntity<GeneralResponse<?>> createShipment(
                        @RequestBody ShipmentCreationForm form) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @PatchMapping("/update/{shipmentId}")
        public ResponseEntity<GeneralResponse<?>> updateShipment(
                        @RequestBody ShipmentUpdateForm form,
                        @PathVariable String shipmentId) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @PatchMapping("/manage/{shipmentId}/import")
        public ResponseEntity<GeneralResponse<?>> importShipment(
                        @PathVariable String shipmentId) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @DeleteMapping("/delete/{shipmentId}")
        public ResponseEntity<GeneralResponse<?>> deleteShipment(
                        @PathVariable String shipmentId) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @PostMapping("/create/{shipmentId}/add")
        public ResponseEntity<GeneralResponse<?>> addItemToShipment(
                        @PathVariable String shipmentId,
                        @RequestBody ShipmentItemAddForm form) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @PatchMapping("/update/{shipmentId}/update/{itemId}")
        public ResponseEntity<GeneralResponse<?>> updateItemInShipment(
                        @PathVariable String shipmentId,
                        @PathVariable String itemId,
                        @RequestBody InventoryItemDetailUpdateForm itemForm) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @PatchMapping("/update/{shipmentId}/delete/{itemId}")
        public ResponseEntity<GeneralResponse<?>> deleteItemFromShipment(
                        @PathVariable String shipmentId,
                        @PathVariable String itemId) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @PatchMapping("/update/{shipmentId}/facility")
        public ResponseEntity<GeneralResponse<?>> updateShipmentFacility(
                        @PathVariable String shipmentId,
                        @RequestBody ShipmentFacilityUpdateForm form) {
                throw new AppException(AppErrorCode.UNIMPLEMENTED);
        }

        @GetMapping("/search/{shipmentId}")
        public ResponseEntity<GeneralResponse<ShipmentDetailResponse>> getShipmentDetail(
                        @PathVariable String shipmentId) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[ShipmentController] Get shipment detail {} for tenantId {}", shipmentId, tenantId);
                ShipmentEntity shipment = shipmentService.getShipment(shipmentId, tenantId);
                List<InventoryItemDetailEntity> items = inventoryItemDetailService.getItemsByShipmentId(shipmentId,
                                tenantId);
                ShipmentDetailResponse response = ShipmentDetailResponse.fromEntity(shipment, items);
                return ResponseEntity.ok(GeneralResponse.success("Successfully get shipment detail", response));
        }

        @GetMapping("/search/by-order/{orderId}")
        public ResponseEntity<GeneralResponse<List<ShipmentEntity>>> getShipmentsByOrderId(
                        @PathVariable String orderId) {
                Long tenantId = authUtils.getCurrentTenantId()
                                .orElseThrow(() -> new AppException(AppErrorCode.UNAUTHORIZED));
                log.info("[ShipmentController] Get shipments by order ID {} for tenantId {}", orderId, tenantId);
                List<ShipmentEntity> shipments = shipmentService.findByOrderId(orderId, tenantId);
                return ResponseEntity.ok(GeneralResponse.success("Successfully get shipments by order ID", shipments));
        }

}
